Plutarch/Rework.hs

```diff
@@ -0,0 +1,491 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unused-imports #-}
+{-# OPTIONS_GHC -Wno-unused-local-binds #-}
+-- This needs to go later
+{-# OPTIONS_GHC -Wwarn #-}
+
+module Plutarch.Rework where
+
+import Generics.SOP
+
+import Generics.SOP qualified as SOP
+
+import GHC.TypeLits
+
+import Data.Functor.Compose
+
+import Control.Monad.Reader
+import Data.Kind (Type)
+import Data.List (groupBy, sortBy)
+import Data.Proxy
+import Data.Text qualified as Text
+import Data.Void
+import GHC.Generics qualified as GHC
+import Plutarch
+import Plutarch.Builtin
+import Plutarch.DataRepr.Internal
+import Plutarch.Internal
+import Plutarch.Prelude
+import Plutarch.Unsafe
+
+newtype PStruct (struct :: [[S -> Type]]) (s :: S) = PStruct {unPStruct :: SOP (Term s) struct}
+newtype PRec (struct :: [S -> Type]) (s :: S) = PRec {unPRec :: NP (Term s) struct}
+
+----------------------------------------------------------------------Smart plet
+pdataStructureAsData :: Term s (PDataStruct struct) -> Term s PData
+pdataStructureAsData = punsafeCoerce
+
+getRawTerm :: Term s a -> TermCont s RawTerm
+getRawTerm t = TermCont $ \f ->
+  Term $ \i ->
+    case runReaderT (runTermMonad (asRawTerm t i)) mempty of
+      Right rt -> asRawTerm (f $ getTerm rt) i
+      Left e -> asRawTerm (pthrow e) i
+
+findOccurrence :: Dig -> Term s a -> TermCont s Int
+findOccurrence targetHash t = TermCont $ \f -> unTermCont $ do
+  rt <- getRawTerm t
+  let
+    go (RLamAbs _ rt) = go rt
+    go (RApply rt rts) = go rt + sum (go <$> rts)
+    go (RForce rt) = go rt
+    go (RDelay rt) = go rt
+    go (RPlaceHolder hash) = if hash == targetHash then 1 else 0
+    go _ = 0
+
+  pure $ f (go rt)
+
+hasErrorTerm :: Term s a -> TermCont s Bool
+hasErrorTerm t = TermCont $ \f -> unTermCont $ do
+  rt <- getRawTerm t
+  let
+    go (RLamAbs _ rt) = go rt
+    go (RApply rt rts) = go rt || any go rts
+    go (RForce rt) = go rt
+    go (RDelay rt) = go rt
+    go (RHoisted (HoistedTerm _ rt)) = go rt
+    go RError = True
+    go _ = False
+
+  pure $ f (go rt)
+
+createPlaceholder :: Term s a -> TermCont s (Dig, Term s a)
+createPlaceholder t = do
+  (hash :: Dig) <- hashOpenTerm t
+  pure (hash, Term $ const $ pure $ TermResult (RPlaceHolder hash) [])
+
+pletSmart :: Term s a -> (Term s a -> Term s b) -> Term s b
```

Comment:
Does this account for stuff like

```
let x = ... in map (\_ -> x) list
```

Note that UPLC is strict.

Reply:
yeah, it won't be able to figure out if bound term will be used multiple times or not

Reply:
The solution is to cover a small set of cases that are useful and happen in practice and are easy enough to detect.
You can replace it with checking whether the bound variable ever appears under a `delay` or `lam`. If it does not, that should mean it's always only evaluated exactly once, hence you can inline it without changing _almost_ anything.

Three properties to uphold:
- Algorithmic complexities must not change
- What errs without optimizations must not err with optimizations.
- What succeeds without optimizations must succeed with optimizations.

The bottom two are a very unfortunate side effect of UPLC's design. I made a CIP <https://github.com/cardano-foundation/CIPs/pull/469> to fix this, but it wasn't very popular.

Reply:
The topmost one wouldn't be affect by doing laziness like Haskell though, since sharing doesn't happen under lambdas. It does though on HVM and stuff like that AFAIU.